Fix MethodInfo Emit for generic interfaces

wzchua opened the pull request:
Signature generation now includes CustomModifiers
MethodBuilder and ConstructorBuilder passes their SignatureHelper for
generation via GetMethodSignature()

Based on dotnet/coreclr#17881

Fixes #25958

wzchua committed:
Fixed Emit MethodInfo for interface generic

Signature generation now includes CustomModifiers
MethodBuilder and ConstructorBuilder passes their SignatureHelper for
generation

Fixes #25958

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=40587) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=40587) before we can accept your contribution.<br/><br/>:x: wzchua [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=40587)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=40587) it.</sub>

steveharter commented:
Mono tests are failing:
```
    System.Reflection.Emit.Tests.ILGeneratorEmitMethodInfo.EmitMethodInfo [FAIL]
      System.InvalidProgramException : Invalid IL code in DynamicType:Call (System.Reflection.Emit.Tests.ILGeneratorEmitMethodInfo/IWithIn`1<int>): IL_0006: callvirt  0x0a000001
      
      
      Stack Trace:
        /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(384,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

wzchua commented:
Hmm, it's failing the test I added. But I'm not sure how to go about debugging the reflection on mono

steveharter commented:
@marek-safar can you provide pointers to wzchua on Mono? Thanks

steveharter commented on src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs:
```diff
@@ -387,55 +386,22 @@ private int GetMemberRefToken(MethodBase method, IEnumerable<Type>? optionalPara
+                sigHelp = GetMemberRefSignature(method, cGenericParameters);
+            }
+
+            if (optionalParameterTypes is {} && optionalParameterTypes.Length > 0)
```
I believe you can change this to `if (optionalParameterTypes?.Length)`

steveharter commented on src/libraries/System.Reflection.Emit.ILGeneration/tests/ILGenerator/EmitMethodInfo.cs:
```diff
@@ -0,0 +1,60 @@
+
+        public sealed class WithIn : IWithIn<int>
+        {
+            public void Method(in RuntimeMethodHandle arg)
```
nit: this syntax `public void Method(in RuntimeMethodHandle arg) { }` is a bit more common.

steveharter commented on src/libraries/System.Reflection.Emit.ILGeneration/tests/ILGenerator/EmitMethodInfo.cs:
```diff
@@ -0,0 +1,60 @@
+
+            ModuleBuilder moduleBuilder = Helpers.DynamicModule(); ;
+            var typeBuilder = moduleBuilder.DefineType("DynamicType", TypeAttributes.Public);
+
```
What happens if there are existing calls to `AddInterfaceImplementation` for example?

i.e. is this a breaking change?

steveharter commented:
@wzchua to skip the test on Mono, your can add this to your test method:
`[SkipOnMono("https://github.com/dotnet/runtime/issues/<AddNewIssueHere")]`

This requires a new issue with a "runtime-mono" label.

steveharter commented:
Since the 5.0 release has no more preview releases, and this is not a risk-free issue, we should consider committing this after we branch for 6.0 which is occurring August 17th.

After it is in master, it is possible to merge this into the 5.0 branch (through an ask mode policy) or a future servicing branch but only if we are confident that it is not breaking and is blocking scenarios.

wzchua committed:
Fixed EmitMethodInfo test case

mono has a per domain scope for methodhandles
Updated code as per review
removed while loop for explicit instance check

wzchua commented:
My test was wrong and I have fixed it

jkotas commented on src/libraries/System.Reflection.Emit.ILGeneration/tests/ILGenerator/EmitMethodInfo.cs:
```diff
@@ -0,0 +1,48 @@
+        public void EmitMethodInfo()
+        {
+            var methodType = typeof(IWithIn<int>);
+            var method = methodType.GetMethod("Method");
```
Nit: We use `var` only when the type is obvious from the right side.

https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md

jkotas reviewed (approved):
Looks good for .NET 6

wzchua committed:
Fixed var styling issue

wzchua committed:
Fixed the other vars missed

steveharter commented:
Test failures appear unrelated: one timeout and one OSX failure in `System.Net.WebSockets.Client.Tests.ConnectTest.ConnectAsync_CancellationRequestedAfterConnect_ThrowsOperationCanceledException`

stakx commented:
@wzchua, I'm glad to see this long-standing issue finally resolved. Thanks for working on it!
